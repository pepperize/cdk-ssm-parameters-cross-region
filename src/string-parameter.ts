import { ContextProvider, RemovalPolicy, Stack, TagManager, TagType, Tokenization } from "aws-cdk-lib";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { IStringParameter } from "aws-cdk-lib/aws-ssm";
import * as cxschema from "aws-cdk-lib/cloud-assembly-schema";
import * as custom_resources from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";
import { ParameterBase, ParameterOptions } from "./parameter-base";
import { Validators } from "./validators";

/**
 * Properties needed to create a String SSM parameter.
 */
export interface StringParameterProps extends ParameterOptions {
  /**
   * The name of the parameter.
   *
   * It may not be a
   *
   * @default - a name will be generated by CloudFormation
   * @stability stable
   */
  readonly parameterName: string;

  /**
   * The region to create the parameter in. See AWS.SSM.region for more information.
   */
  readonly region: string;

  /**
   * The value of the parameter.
   *
   * It may not reference another parameter and ``{{}}`` cannot be used in the value.
   */
  readonly stringValue: string;

  /**
   * The data type of the parameter, such as `text` or `aws:ec2:image`.
   *
   * @default - undefined
   */
  readonly dataType?: ssm.ParameterDataType;

  /**
   * Whether to retain or delete the parameter on CloudFormation delete event.
   *
   * @default - DESTROY
   */
  readonly removalPolicy?: RemovalPolicy;
}

export interface StringParameterAttributes extends ssm.StringParameterAttributes {
  /**
   * The region to retrieve the parameter from. See AWS.SSM.region for more information.
   */
  readonly region: string;
}

const RESOURCE_TYPE = "Custom::SSM_String_Parameter_Cross_Region";

export class StringParameter extends ParameterBase implements ssm.IStringParameter {
  /**
   * Imports an external string parameter with name and optional version.
   */
  public static fromStringParameterAttributes(
    scope: Construct,
    id: string,
    attrs: StringParameterAttributes
  ): ssm.IStringParameter {
    if (!attrs.parameterName) {
      throw new Error("parameterName cannot be an empty string");
    }

    const parameterType = attrs.type ?? ssm.ParameterType.STRING;

    const parameterName = attrs.version
      ? `${attrs.parameterName}:${Tokenization.stringifyNumber(attrs.version)}`
      : attrs.parameterName;

    class Import extends ParameterBase {
      public readonly parameterName: string;
      public readonly parameterArn: string;
      public readonly parameterType: ssm.ParameterType;
      public readonly stringValue: string;

      constructor() {
        super(scope, id);

        this.parameterArn = Stack.of(this).formatArn({
          service: "ssm",
          resource: `parameter${parameterName}`,
          region: attrs.region,
        });

        const parameter = new custom_resources.AwsCustomResource(this, "Resource", {
          resourceType: RESOURCE_TYPE,
          onCreate: {
            region: attrs.region,
            service: "SSM",
            action: "getParameter", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html#getParameter-property
            parameters: {
              Name: parameterName,
            },
            physicalResourceId: custom_resources.PhysicalResourceId.of(this.parameterArn),
          },
          onUpdate: {
            region: attrs.region,
            service: "SSM",
            action: "getParameter", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html#getParameter-property
            parameters: {
              Name: parameterName,
            },
            physicalResourceId: custom_resources.PhysicalResourceId.of(this.parameterArn),
          },
          policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
            resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
          }),
        });

        this.parameterName = attrs.parameterName;
        this.parameterType = parameterType;
        this.stringValue = parameter.getResponseField("Parameter.Value");
      }
    }

    return new Import();
  }

  /**
   * Imports an external string parameter by name and region.
   */
  public static fromStringParameterName(
    scope: Construct,
    id: string,
    region: string,
    parameterName: string
  ): IStringParameter {
    return this.fromStringParameterAttributes(scope, id, { parameterName: parameterName, region: region });
  }

  /**
   * Reads the value of an SSM parameter during synthesis through an
   * environmental context provider.
   *
   * Requires that the stack this scope is defined in will have explicit
   * account information. Otherwise, it will fail during synthesis.
   */
  public static valueFromLookup(scope: Construct, region: string, parameterName: string): string {
    return ContextProvider.getValue(scope, {
      provider: cxschema.ContextProvider.SSM_PARAMETER_PROVIDER,
      props: { parameterName: parameterName, region: region },
      dummyValue: `dummy-value-for-${parameterName}`,
    }).value;
  }

  public readonly parameterArn: string;
  public readonly parameterName: string;
  public readonly parameterType: string;
  public readonly stringValue: string;

  readonly tags = new TagManager(TagType.KEY_VALUE, RESOURCE_TYPE);

  constructor(scope: Construct, id: string, props: StringParameterProps) {
    super(scope, id, {
      physicalName: props.parameterName,
    });

    if (props.allowedPattern) {
      Validators.parameterValue(props.stringValue, props.allowedPattern);
    }
    Validators.parameterName(props.parameterName);
    Validators.description(props.description);

    this.parameterArn = Stack.of(this).formatArn({
      service: "ssm",
      resource: `parameter${props.parameterName}`,
      region: props.region,
    });

    const removalPolicy = props.removalPolicy ?? RemovalPolicy.DESTROY;

    new custom_resources.AwsCustomResource(this, "Resource", {
      resourceType: RESOURCE_TYPE,
      onCreate: {
        region: props.region,
        service: "SSM",
        action: "putParameter", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html#putParameter-property
        parameters: {
          Name: props.parameterName,
          Value: props.stringValue,
          AllowedPattern: props.allowedPattern,
          DataType: props.dataType,
          Description: props.description,
          Tags: this.tags.renderedTags,
          Tier: props.tier,
          Type: ssm.ParameterType.STRING,
        },
        physicalResourceId: custom_resources.PhysicalResourceId.of(this.parameterArn),
      },
      onUpdate: {
        region: props.region,
        service: "SSM",
        action: "putParameter", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html#putParameter-property
        parameters: {
          Name: props.parameterName,
          Value: props.stringValue,
          AllowedPattern: props.allowedPattern,
          DataType: props.dataType,
          Description: props.description,
          Tags: this.tags.renderedTags,
          Tier: props.tier,
          Type: ssm.ParameterType.STRING,
          Overwrite: true,
        },
        physicalResourceId: custom_resources.PhysicalResourceId.of(this.parameterArn),
      },
      onDelete:
        removalPolicy != RemovalPolicy.RETAIN
          ? {
              region: props.region,
              service: "SSM",
              action: "deleteParameter", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html#deleteParameter-property
              parameters: {
                Name: props.parameterName,
              },
              physicalResourceId: custom_resources.PhysicalResourceId.of(this.parameterArn),
            }
          : undefined,
      policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
        resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    this.parameterName = props.parameterName;
    this.parameterType = ssm.ParameterType.STRING;
    this.stringValue = props.stringValue;
  }
}
