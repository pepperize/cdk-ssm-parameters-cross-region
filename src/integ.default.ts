import { App, Environment, Stack } from "aws-cdk-lib";
import { StringParameter } from "./string-parameter";

// export CDK_DEFAULT_REGION=us-east-1
// export CDK_DEFAULT_ACCOUNT=123456789012
// cdk deploy --app 'npx ts-node -P tsconfig.json --prefer-ts-exts ./src/integ.default.ts'

const app = new App();

const env: Environment = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new Stack(app, "SsmParameterCrossRegionIntegrationTest", { env: env });

const putParameter = new StringParameter(stack, "PutParameter", {
  region: "eu-central-1",
  parameterName: "/path/name/integ/test",
  stringValue: "Say hello from another region",
});
const getParameter = StringParameter.fromStringParameterName(
  stack,
  "GetParameter",
  "eu-central-1",
  "/path/name/integ/test"
);
getParameter.node.addDependency(putParameter);

const arn = stack.formatArn({
  service: "ssm",
  resource: `parameter/path/name/integ/test`,
  region: "eu-central-1",
  account: env.account,
});
const getByArn = StringParameter.fromStringParameterArn(stack, "GetParameterByArn", arn);
getByArn.node.addDependency(putParameter);

app.synth();
