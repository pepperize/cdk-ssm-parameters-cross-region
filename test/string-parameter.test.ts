import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { StringParameter } from "../src";

describe("StringParameter", () => {
  it("Should create and import parameter", () => {
    // Given
    const stack = new Stack();

    // When
    const putParameter = new StringParameter(stack, "PutParameter", {
      region: "eu-central-1",
      parameterName: "/path/name/integ/test",
      stringValue: "Say hello from another region",
    });
    const getParameter = StringParameter.fromStringParameterName(
      stack,
      "GetParameter",
      "eu-central-1",
      "/path/name/integ/test"
    );
    getParameter.node.addDependency(putParameter);

    // When import by ARN
    const arn = `arn:aws:ssm:eu-central-1:123456789012:parameter/path/name/integ/test`;
    const getByArn = StringParameter.fromStringParameterArn(stack, "GetByArn", arn);
    getByArn.node.addDependency(putParameter);

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});
