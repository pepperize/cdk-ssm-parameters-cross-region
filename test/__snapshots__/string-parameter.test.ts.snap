// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringParameter Should create and import parameter 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetByArnCustomResourcePolicy3524B48B": Object {
      "DependsOn": Array [
        "PutParameterCustomResourcePolicy3B520E2B",
        "PutParameter9BA39D58",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetByArnCustomResourcePolicy3524B48B",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetByArnD4C43426": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "GetByArnCustomResourcePolicy3524B48B",
        "PutParameterCustomResourcePolicy3B520E2B",
        "PutParameter9BA39D58",
      ],
      "Properties": Object {
        "Create": "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"arn:aws:ssm:eu-central-1:123456789012:parameter/path/name/integ/test\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:aws:ssm:eu-central-1:123456789012:parameter/path/name/integ/test\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"arn:aws:ssm:eu-central-1:123456789012:parameter/path/name/integ/test\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:aws:ssm:eu-central-1:123456789012:parameter/path/name/integ/test\\"}}",
      },
      "Type": "Custom::SSM_String_Parameter_Cross_Region",
      "UpdateReplacePolicy": "Delete",
    },
    "GetParameterCustomResourcePolicyDE2625D5": Object {
      "DependsOn": Array [
        "PutParameterCustomResourcePolicy3B520E2B",
        "PutParameter9BA39D58",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetParameterCustomResourcePolicyDE2625D5",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetParameterFEF96F3D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "GetParameterCustomResourcePolicyDE2625D5",
        "PutParameterCustomResourcePolicy3B520E2B",
        "PutParameter9BA39D58",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/path/name/integ/test\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":ssm:eu-central-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":parameter/path/name/integ/test\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"getParameter\\",\\"parameters\\":{\\"Name\\":\\"/path/name/integ/test\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":ssm:eu-central-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":parameter/path/name/integ/test\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::SSM_String_Parameter_Cross_Region",
      "UpdateReplacePolicy": "Delete",
    },
    "PutParameter9BA39D58": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "PutParameterCustomResourcePolicy3B520E2B",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/path/name/integ/test\\",\\"Value\\":\\"Say hello from another region\\",\\"Type\\":\\"String\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":ssm:eu-central-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":parameter/path/name/integ/test\\"}}",
            ],
          ],
        },
        "Delete": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"deleteParameter\\",\\"parameters\\":{\\"Name\\":\\"/path/name/integ/test\\"},\\"physicalResourceId\\":{\\"id\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":ssm:eu-central-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":parameter/path/name/integ/test\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"region\\":\\"eu-central-1\\",\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/path/name/integ/test\\",\\"Value\\":\\"Say hello from another region\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"physicalResourceId\\":{\\"id\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":ssm:eu-central-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":parameter/path/name/integ/test\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::SSM_String_Parameter_Cross_Region",
      "UpdateReplacePolicy": "Delete",
    },
    "PutParameterCustomResourcePolicy3B520E2B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ssm:DeleteParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PutParameterCustomResourcePolicy3B520E2B",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
